### PR TITLE
Implement a BitField hint

### PR DESCRIPTION
**Pull request by reduz**
Allows to specify the binder that an enum must be treated as a bitfield. From the Godot side this does not change much, but it allows on the binder side to discern when something is meant to be used as enum and when its meant to be used as flags.

This is a common problem for languages such as C# or Rust. Additionally, Godot users can often feel confused on how to use bitfields, since they did not necesarily study this aspect of programming, so a dedicated API like this would improve on ease of use.

Usage:

For enums used as bitfields (flags) instead of: 
```C++
VARIANT_ENUM_CAST(Class::Enum)
BIND_ENUM_CONSTANT(FLAG_NAME)
void SomeClass::do_something(uint32_t p_flags)
uint32_t SomeClass::get_flags() const;
//or, sometimes, though incorrect for bitmasks:
void SomeClass::do_something(Enum p_flags)
Enum SomeClass::get_flags() const;
```

which is currently used for enums used as bitfields, 

you must now use:
```C++
VARIANT_BITFIELD_CAST(Class::Enum)
BIND_BITFIELD_FLAG(FLAG_NAME)
void SomeClass::do_something(BitField<Enum> p_value) 
BitField<Enum> SomeClass::get_flags() const;
```

The `BitField<>` template also has some functions to ensure it only takes flags from the proper enum type.

For GDEXtension users, an extra field is added to enums at the JSON level:
`"is_bitfield": true`

Flags are also detailed as such in the built-in doc.
image

Supersedes #61738

TODO (beyond this PR): Currently, what is missing is replacing all usages of enums as flags for bitfields. Also for doc generators based on the XML files, parsing the new bitfield hint is pending:

### CODE REVIEW

**Comment by akien-mga** [comment]
There's another `enum_map` usage to change in Mono (same usage as in the ClassDB test, should be easy to do even without compiling locally):
```
modules/mono/editor/bindings_generator.cpp: In member function 'bool BindingsGenerator::_populate_object_type_interfaces()':
modules/mono/editor/bindings_generator.cpp:3078:71: error: invalid initialization of reference of type 'const HashMap<StringName, List<StringName> >&' from expression of type 'HashMap<StringName, ClassDB::ClassInfo::EnumInfo>'
 3078 |   const HashMap<StringName, List<StringName>> &enum_map = class_info->enum_map;
      |                                                           ~~~~~~~~~~~~^~~~~~~~
```

> Also for doc generators based on the XML files, parsing the new bitfield hint is pending:

This would need to be done in this PR too so that it can pass CI, as `make_rst.py` is used for validation of the XML reference. Maybe @YuriSizov or @Calinou can provide a diff, it shouldn't be difficult to add.

**Comment by reduz** [comment]
@akien-mga Thanks! Just fixed it, it seems I forgot to compile with Mono enabled.

**Comment by raulsntos** [comment]
This PR does not seem to support bitfield enums in core constants (e.g.: `PROPERTY_USAGE_*`), is there a reason for that or was it an oversight?

---

Also, these changes to C# bindings generator should fix the CI:

```diff
diff --git a/modules/mono/editor/bindings_generator.cpp b/modules/mono/editor/bindings_generator.cpp
index f29e29f2ff..4ee774c3bd 100644
--- a/modules/mono/editor/bindings_generator.cpp
+++ b/modules/mono/editor/bindings_generator.cpp
@@ -2865,7 +2865,7 @@ bool BindingsGenerator::_populate_object_type_interfaces() {
 							   " We only expected Object.free, but found '" +
 							itype.name + "." + imethod.name + "'.");
 				}
-			} else if (return_info.type == Variant::INT && return_info.usage & PROPERTY_USAGE_CLASS_IS_ENUM) {
+			} else if (return_info.type == Variant::INT && return_info.usage & (PROPERTY_USAGE_CLASS_IS_ENUM | PROPERTY_USAGE_CLASS_IS_BITFIELD)) {
 				imethod.return_type.cname = return_info.class_name;
 				imethod.return_type.is_enum = true;
 			} else if (return_info.class_name != StringName()) {
@@ -2903,7 +2903,7 @@ bool BindingsGenerator::_populate_object_type_interfaces() {
 				ArgumentInterface iarg;
 				iarg.name = orig_arg_name;
 
-				if (arginfo.type == Variant::INT && arginfo.usage & PROPERTY_USAGE_CLASS_IS_ENUM) {
+				if (arginfo.type == Variant::INT && arginfo.usage & (PROPERTY_USAGE_CLASS_IS_ENUM | PROPERTY_USAGE_CLASS_IS_BITFIELD)) {
 					iarg.type.cname = arginfo.class_name;
 					iarg.type.is_enum = true;
 				} else if (arginfo.class_name != StringName()) {
@@ -3011,7 +3011,7 @@ bool BindingsGenerator::_populate_object_type_interfaces() {
 				ArgumentInterface iarg;
 				iarg.name = orig_arg_name;
 
-				if (arginfo.type == Variant::INT && arginfo.usage & PROPERTY_USAGE_CLASS_IS_ENUM) {
+				if (arginfo.type == Variant::INT && arginfo.usage & (PROPERTY_USAGE_CLASS_IS_ENUM | PROPERTY_USAGE_CLASS_IS_BITFIELD)) {
 					iarg.type.cname = arginfo.class_name;
 					iarg.type.is_enum = true;
 				} else if (arginfo.class_name != StringName()) {
```

**Comment by reduz** [comment]
@raulsntos I am not sure if there is any, does not seem like its the case

**Comment by raulsntos** [comment]
These core enums have their values defined as power of two and/or they are used in bitwise operations throughout the engine:

- `InlineAlignment` (`INLINE_ALIGNMENT_*`)
- `KeyModifierMask` (`KEY_MASK_*`)
- `PropertyUsageFlags` (`PROPERTY_USAGE_*`)
- `MethodFlags` (`METHOD_FLAG_*`)

**Comment by reduz** [comment]
@raulsntos oh good point about this, I am not entirely sure if they are used as bitfields anywhere though, but I can add support either way.

File: core/variant/type_info.h [hunk 1]
```diff
@@ -279,6 +279,51 @@ inline StringName __constant_get_enum_name(T param, const String &p_constant) {
...
+template <class T>
+class BitField {
+	uint32_t value = 0;
+
+public:
+	_FORCE_INLINE_ void set_flag(T p_flag) { value |= p_flag; }
+	_FORCE_INLINE_ bool has_flag(T p_flag) const { return value & p_flag; }
+	_FORCE_INLINE_ void clear_flag(T p_flag) { return value &= ~p_flag; }
+	_FORCE_INLINE_ BitField(uint32_t p_value) { value = p_value; }
+	_FORCE_INLINE_ operator uint32_t() const { return value; }
```
Comment (AndreaCatania) [inline core/variant/type_info.h:291]: I wonder if `==` is needed, or it's implicitly implemented. So to support this case:
```c++
if( flags == BitField<Enum>(ENUM_1 | ENUM_2 | ENUM_3) )
{
```

**Review by AndreaCatania — APPROVED** [review]
It's a positive improvement! LGTM

**Comment by akien-mga** [comment]
Pushed an update with both @raulsntos and @YuriSizov's patches amended in so it should now pass CI.

**Review by akien-mga — APPROVED** [review]
Looks good.

Let's tackle the remaining core constants mentioned in https://github.com/godotengine/godot/pull/62374#issuecomment-1166321325 in a follow-up.

**Comment by bruvzg** [comment]
Something is probably missing from the binder. I have tried to move TextServer to the new API and `BitField<Enum> SomeClass::get_flags() const;` type method binds always fail to build:

```
[ 83%] In file included from servers/text_server.cpp:31:
In file included from ./servers/text_server.h:34:
In file included from ./core/object/ref_counted.h:34:
In file included from ./core/object/class_db.h:34:
In file included from ./core/object/method_bind.h:34:
./core/variant/binder_common.h:728:8: error: no viable overloaded '='
        r_ret = (p_instance->*p_method)(VariantCasterAndValidate<P>::cast(p_args, Is, r_error)...);
        ~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[ 83%] variant/binder_common.h:523:2: note: in instantiation of function template specialization 'call_with_variant_args_retc_helper<__UnexistingClass, BitField<TextServer::FontStyle>, const RID &, 0UL>' requested here
        call_with_variant_args_retc_helper(p_instance, p_method, args, r_ret, r_error, BuildIndexSequence<sizeof...(P)>{});
        ^
[ 83%] ./core/object/method_bind.h:568:3: note: in instantiation of function template specialization 'call_with_variant_args_retc_dv<__UnexistingClass, BitField<TextServer::FontStyle>, const RID &>' requested here
                call_with_variant_args_retc_dv(reinterpret_cast<MB_T *>(p_object), method, p_args, p_arg_count, ret, r_error, get_default_arguments());
                ^
./core/object/method_bind.h:581:2: note: in instantiation of member function 'MethodBindTRC<BitField<TextServer::FontStyle>, const RID &>::call' requested here
        MethodBindTRC(R (MB_T::*p_method)(P...) const) {
        ^
./core/object/method_bind.h:595:26: note: in instantiation of member function 'MethodBindTRC<BitField<TextServer::FontStyle>, const RID &>::MethodBindTRC' requested here
        MethodBind *a = memnew((MethodBindTRC<R, P...>)(reinterpret_cast<R (MB_T::*)(P...) const>(p_method)));
                                ^
./core/object/class_db.h:250:22: note: in instantiation of function template specialization 'create_method_bind<TextServer, BitField<TextServer::FontStyle>, const RID &>' requested here
                MethodBind *bind = create_method_bind(p_method);
                                   ^
servers/text_server.cpp:217:11: note: in instantiation of function template specialization 'ClassDB::bind_method<MethodDefinition, BitField<TextServer::FontStyle> (TextServer::*)(const RID &) const>' requested here
        ClassDB::bind_method(D_METHOD("font_get_style", "font_rid"), &TextServer::font_get_style);
                 ^
./core/variant/variant.h:731:7: note: candidate function not viable: no known conversion from 'BitField<TextServer::FontStyle>' to 'const Variant' for 1st argument
        void operator=(const Variant &p_variant); // only this is enough for all the other types
             ^
[ 84%] scene/resources/font.cpp:31:
```

**Pull request merged**